add new resource & data source: `azurerm_netapp_account`

neil-yechenwei opened the pull request:
This PR is code implement for issue https://github.com/terraform-providers/terraform-provider-azurerm/issues/4417

jstewart612 commented:
Ping.  Pong?

katbyte commented on website/docs/r/netapp_account.html.markdown:
```diff
@@ -0,0 +1,86 @@
+
+* `name` - (Required) The name of the NetApp Account. Changing this forces a new resource to be created.
+
+* `resource_group_name` - (Required) The resource group name of the NetApp Account. Changing this forces a new resource to be created.
```
Could we word this like the rest of the documentatin?
```suggestion
* `resource_group_name` - (Required) The name of the resource group where the NetApp Account resides. Changing this forces a new resource to be created.
```

neil-yechenwei replied:
Fixed

katbyte commented on website/docs/r/netapp_account.html.markdown:
```diff
@@ -0,0 +1,86 @@
+
+* `active_directories` - (Optional) One or more `active_directories` block defined below.
+
+* `tags` - (Optional) Resource tags. Changing this forces a new resource to be created.
```
Could we match the rest of the docs:
```suggestion
* `tags` - (Optional) A mapping of tags to assign to the resource.
```

neil-yechenwei replied:
Fixed

katbyte commented on website/docs/r/netapp_account.html.markdown:
```diff
@@ -0,0 +1,86 @@
+
+* `smb_server_name` - (Required) NetBIOS name of the SMB server. This name will be registered as a computer account in the AD and used to mount volumes.
+
+* `username` - (Required) Username of Active Directory domain administrator.
```
Is this the ad admin used to add the netapp to the domain? could we clarify the docs here?

neil-yechenwei replied:
Fixed

katbyte commented on website/docs/d/netapp_account.html.markdown:
```diff
@@ -0,0 +1,65 @@
+
+* `active_directories` - One or more `active_directories` block defined below.
+
+* `tags` - Resource tags.
```
Could we reword this to match the rest of the documentation?

neil-yechenwei replied:
Fixed

katbyte commented on website/docs/d/netapp_account.html.markdown:
```diff
@@ -0,0 +1,65 @@
+
+The following attributes are exported:
+
+* `location` - Resource location.
```
Could we reword this to match the rest of the documentation?

neil-yechenwei replied:
Fixed

neil-yechenwei commented on azurerm/data_source_netapp_account.go:
```diff
@@ -0,0 +1,104 @@
+
+			"resource_group_name": azure.SchemaResourceGroupNameForDataSource(),
+
+			"active_directories": {
```
Fixed

neil-yechenwei commented on azurerm/data_source_netapp_account.go:
```diff
@@ -0,0 +1,104 @@
+				},
+			},
+
+			"tags": tagsForDataSourceSchema(),
```
Fixed

neil-yechenwei commented on azurerm/data_source_netapp_account.go:
```diff
@@ -0,0 +1,104 @@
+	if location := resp.Location; location != nil {
+		d.Set("location", azure.NormalizeLocation(*location))
+	}
+	if accountProperties := resp.AccountProperties; accountProperties != nil {
```
Fixed

neil-yechenwei commented on azurerm/data_source_netapp_account.go:
```diff
@@ -0,0 +1,104 @@
+
+			"location": azure.SchemaLocationForDataSource(),
+
+			"resource_group_name": azure.SchemaResourceGroupNameForDataSource(),
```
Fixed

neil-yechenwei commented on azurerm/resource_arm_netapp_account.go:
```diff
@@ -0,0 +1,276 @@
+
+			"location": azure.SchemaLocation(),
+
+			"resource_group_name": azure.SchemaResourceGroupNameForDataSource(),
```
Fixed

katbyte replied:
shouldn't this be
```suggestion
			"resource_group_name": azure.SchemaResourceGroupName(),
```

neil-yechenwei commented on azurerm/resource_arm_netapp_account.go:
```diff
@@ -0,0 +1,276 @@
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"id": {
```
Fixed

neil-yechenwei commented on website/docs/r/netapp_account.html.markdown:
```diff
@@ -0,0 +1,86 @@
+
+The following attributes are exported:
+
+* `id` - Resource id.
```
Fixed.

neil-yechenwei commented on website/docs/r/netapp_account.html.markdown:
````diff
@@ -0,0 +1,86 @@
+NetApp Account can be imported using the `resource id`, e.g.
+
+```shell
+$ terraform import azurerm_netapp_account.example /subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/acctestRG/providers/Microsoft.NetApp/netAppAccounts/
````
Fixed

neil-yechenwei commented on azurerm/resource_arm_netapp_account.go:
```diff
@@ -0,0 +1,276 @@
+							Required:     true,
+							ValidateFunc: validate.NoEmptyStrings,
+						},
+						"status": {
```
Fixed

neil-yechenwei commented on azurerm/resource_arm_netapp_account.go:
```diff
@@ -0,0 +1,276 @@
+							Required:     true,
+							ValidateFunc: validate.NoEmptyStrings,
+						},
+						"organizational_unit": {
```
Fixed

neil-yechenwei commented on azurerm/resource_arm_netapp_account.go:
```diff
@@ -0,0 +1,276 @@
+				return fmt.Errorf("Error checking for present of existing NetApp Account %q (Resource Group %q): %+v", name, resourceGroup, err)
+			}
+		}
+		if !utils.ResponseWasNotFound(resp.Response) {
```
Fixed

neil-yechenwei commented on azurerm/resource_arm_netapp_account.go:
```diff
@@ -0,0 +1,276 @@
+	resourceGroup := d.Get("resource_group_name").(string)
+
+	if features.ShouldResourcesBeImported() && d.IsNewResource() {
+		resp, err := client.Get(ctx, resourceGroup, name)
```
Fixed

neil-yechenwei commented on azurerm/resource_arm_netapp_account.go:
```diff
@@ -0,0 +1,276 @@
+
+			"resource_group_name": azure.SchemaResourceGroupNameForDataSource(),
+
+			"active_directories": {
```
Fixed

neil-yechenwei commented on azurerm/resource_arm_netapp_account.go:
```diff
@@ -0,0 +1,276 @@
+	if location := resp.Location; location != nil {
+		d.Set("location", azure.NormalizeLocation(*location))
+	}
+	if accountProperties := resp.AccountProperties; accountProperties != nil {
```
Fixed

neil-yechenwei commented on azurerm/resource_arm_netapp_account.go:
```diff
@@ -0,0 +1,276 @@
+		userName := v["username"].(string)
+
+		result := netapp.ActiveDirectory{
+			DNS:                utils.String(dns),
```
Fixed

neil-yechenwei commented on azurerm/resource_arm_netapp_account.go:
```diff
@@ -0,0 +1,276 @@
+	}
+
+	for _, item := range *input {
+		v := make(map[string]interface{})
```
Fixed

neil-yechenwei commented on azurerm/resource_arm_netapp_account.go:
```diff
@@ -0,0 +1,276 @@
+		}
+		if organizationalUnit := item.OrganizationalUnit; organizationalUnit != nil {
+			v["organizational_unit"] = *organizationalUnit
+		}
```
Fixed

neil-yechenwei commented on azurerm/resource_arm_netapp_account_test.go:
```diff
@@ -0,0 +1,190 @@
+  }
+
+  tags = {
+    env = "test"
```
Fixed

neil-yechenwei commented on azurerm/resource_arm_netapp_account_test.go:
```diff
@@ -0,0 +1,190 @@
+}
+
+resource "azurerm_netapp_account" "test" {
+  name                = "acctestnetappaccount-%d"
```
Fixed

neil-yechenwei commented on examples/net-app/account/README.md:
```diff
@@ -0,0 +1,3 @@
+## Example: NetApp Account
```
Deleted this file.

neil-yechenwei commented on examples/net-app/account/variables.tf:
```diff
@@ -0,0 +1,7 @@
+
+variable "location" {
+  description = "The Azure location where all resources in this example should be created."
+}
```
Fixed

neil-yechenwei commented on examples/net-app/account/main.tf:
```diff
@@ -0,0 +1,23 @@
+}
+
+resource "azurerm_netapp_account" "test" {
+  name                = "acctestnetappaccount"
```
Fixed

neil-yechenwei commented on examples/net-app/account/main.tf:
```diff
@@ -0,0 +1,23 @@
+  resource_group_name = "${azurerm_resource_group.test.name}"
+
+  active_directories {
+    username            = "aduser"
```
Fixed

neil-yechenwei commented on examples/net-app/account/main.tf:
```diff
@@ -0,0 +1,23 @@
+  }
+
+  tags = {
+    env = "test"
```
Fixed

neil-yechenwei commented on examples/net-app/account/main.tf:
```diff
@@ -0,0 +1,23 @@
+  location = "${var.location}"
+}
+
+resource "azurerm_netapp_account" "test" {
```
Fixed

neil-yechenwei commented on website/docs/d/netapp_account.html.markdown:
```diff
@@ -0,0 +1,65 @@
+
+The `active_directories` block contains the following:
+
+* `id` - Resource ID.
```
Fixed

neil-yechenwei commented on website/docs/d/netapp_account.html.markdown:
```diff
@@ -0,0 +1,65 @@
+
+* `domain` - Name of the Active Directory domain.
+
+* `organizational_unit` - The Organizational Unit (OU) within the Windows Active Directory.
```
Fixed

neil-yechenwei commented on website/docs/d/netapp_account.html.markdown:
```diff
@@ -0,0 +1,65 @@
+
+* `smb_server_name` - NetBIOS name of the SMB server. This name will be registered as a computer account in the AD and used to mount volumes.
+
+* `status` - Status of the Active Directory.
```
It means the provision status of Active Directory like ""status": "Created","

neil-yechenwei commented on website/docs/d/netapp_account.html.markdown:
```diff
@@ -0,0 +1,65 @@
+
+---
+
+The `active_directories` block contains the following:
```
Fixed

neil-yechenwei commented on website/docs/r/netapp_account.html.markdown:
```diff
@@ -0,0 +1,86 @@
+  }
+
+  tags = {
+    env = "test"
```
Fixed

neil-yechenwei commented on website/docs/r/netapp_account.html.markdown:
```diff
@@ -0,0 +1,86 @@
+
+* `resource_group_name` - (Required) The resource group name of the NetApp Account. Changing this forces a new resource to be created.
+
+* `location` - (Required) Resource location. Changing this forces a new resource to be created.
```
Fixed

neil-yechenwei commented on azurerm/data_source_netapp_account.go:
```diff
@@ -0,0 +1,104 @@
+		}
+	}
+
+	return tags.FlattenAndSetTags(d, resp.Tags)
```
Fixed

neil-yechenwei commented on azurerm/internal/tags/flatten.go:
```diff
@@ -29,3 +30,19 @@ func FlattenAndSet(d *schema.ResourceData, tagMap map[string]*string) error {
 	return nil
 }
+
+func FlattenAndSetTags(d *schema.ResourceData, tags interface{}) error {
```
Removed

neil-yechenwei commented on azurerm/data_source_netapp_account.go:
```diff
@@ -0,0 +1,104 @@
+			"name": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: validate.NoEmptyStrings,
```
Fixed

neil-yechenwei commented on azurerm/resource_arm_netapp_account.go:
```diff
@@ -0,0 +1,276 @@
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validate.NoEmptyStrings,
```
Fixed

neil-yechenwei commented on azurerm/resource_arm_netapp_account.go:
```diff
@@ -0,0 +1,276 @@
+						"username": {
+							Type:         schema.TypeString,
+							Required:     true,
+							ValidateFunc: validate.NoEmptyStrings,
```
After checked on azure portal, this username field can accept any string value so that we just need to validate whether it's empty value here

neil-yechenwei commented on azurerm/resource_arm_netapp_account.go:
```diff
@@ -0,0 +1,276 @@
+						"domain": {
+							Type:         schema.TypeString,
+							Required:     true,
+							ValidateFunc: validate.NoEmptyStrings,
```
Fixed

neil-yechenwei commented on azurerm/resource_arm_netapp_account.go:
```diff
@@ -0,0 +1,276 @@
+							Computed: true,
+						},
+						"dns": {
+							Type:         schema.TypeString,
```
Fixed

neil-yechenwei commented on azurerm/resource_arm_netapp_account.go:
```diff
@@ -0,0 +1,276 @@
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"dns": {
```
Fixed

neil-yechenwei commented on website/docs/r/netapp_account.html.markdown:
```diff
@@ -0,0 +1,86 @@
+
+The `active_directories` block supports the following:
+
+* `dns` - (Required) Comma separated list of DNS server IP addresses for the Active Directory domain.
```
Fixed

neil-yechenwei commented on azurerm/resource_arm_netapp_account.go:
```diff
@@ -0,0 +1,276 @@
+						"dns": {
+							Type:         schema.TypeString,
+							Required:     true,
+							ValidateFunc: validate.NoEmptyStrings,
```
Fixed

neil-yechenwei commented:
Hi @katbyte 

Thank you for the comments, I've updated code. Please help to review.

katbyte commented on azurerm/data_source_netapp_account.go:
```diff
@@ -0,0 +1,108 @@
+		}
+	}
+
+	return tags.FlattenAndSet(d, currentTags)
```
I think this can be
```suggestion
	return tags.FlattenAndSet(d, resp.Tags)
```

neil-yechenwei replied:
It seems that we cannot do this since the type of resp.Tags is "Tags interface{} json:"tags,omitempty"" and function tags.FlattenAndSet only can accept type "tagMap map[string]*string". If I write code as "return tags.FlattenAndSet(d, resp.Tags)" and it would throw below error message. So we need this.
![image](https://user-images.githubusercontent.com/19754191/66373739-ecb67180-e9db-11e9-936b-b6b5f922e4ba.png)

katbyte replied:
thanks for the clarification! makes sense

katbyte commented on azurerm/data_source_netapp_account.go:
```diff
@@ -0,0 +1,108 @@
+		for k, v := range tagMap {
+			currentTags[k] = utils.String(v.(string))
+		}
+	}
```
I don't think you need this?

neil-yechenwei replied:
As above explain, so I think we need this.

katbyte commented on azurerm/resource_arm_netapp_account.go:
```diff
@@ -0,0 +1,288 @@
+	return results
+}
+
+func flattenActiveDirectoryDNSServers(input string) []interface{} {
```
You should be able to do `utils.FlattenStringSlice(strings.Split(input, ","))` instead of this bespoke function?

neil-yechenwei replied:
Fixed

katbyte commented on azurerm/resource_arm_netapp_account_test.go:
```diff
@@ -0,0 +1,190 @@
+func testAccAzureRMNetAppAccount_basic(rInt int, location string) string {
+	return fmt.Sprintf(`
+resource "azurerm_resource_group" "test" {
+  name     = "acctestRG-%d"
```
trying something new, could we use 
```suggestion
  name     = "acctestRG-netapp-%d"
```

neil-yechenwei replied:
Fixed

katbyte commented on azurerm/resource_arm_netapp_account_test.go:
```diff
@@ -0,0 +1,190 @@
+func testAccAzureRMNetAppAccount_complete(rInt int, location string) string {
+	return fmt.Sprintf(`
+resource "azurerm_resource_group" "test" {
+  name     = "acctestRG-%d"
```
trying something new, could we use 
```suggestion
  name     = "acctestRG-netapp-%d"
```

neil-yechenwei replied:
Fixed

katbyte commented on website/docs/r/netapp_account.html.markdown:
```diff
@@ -0,0 +1,96 @@
+page_title: "Azure Resource Manager: azurerm_netapp_account"
+sidebar_current: "docs-azurerm-resource-netapp-account"
+description: |-
+  Manage Azure NetApp Account instance.
```
consistency
```suggestion
  Manages Azure NetApp Account instance.
```

neil-yechenwei replied:
Fixed

katbyte commented on website/docs/r/netapp_account.html.markdown:
```diff
@@ -0,0 +1,96 @@
+
+# azurerm_netapp_account
+
+Manage Azure NetApp Account instance.
```
consistency
```suggestion
Manages Azure NetApp Account instance.
```

neil-yechenwei replied:
Fixed

katbyte commented on website/docs/r/netapp_account.html.markdown:
```diff
@@ -0,0 +1,96 @@
+
+The following attributes are exported:
+
+* `id` - Resource id.
```
Could we expand the description here a bit more:
```suggestion
* `id` - The ID of the NetApp resource.
```

neil-yechenwei replied:
FIxed

katbyte commented on website/docs/r/netapp_account.html.markdown:
```diff
@@ -0,0 +1,96 @@
+
+* `smb_server_name` - (Required) NetBIOS name of the SMB server. This name will be registered as a computer account in the AD and used to mount volumes.
+
+* `username` - (Required) Username of Active Directory domain administrator, which have permissions to create SMB server machine account in the AD domain.
```
```suggestion
* `username` - (Required) Username of Active Directory domain administrator, which have permissions to create a SMB machine account in the AD domain.
```

neil-yechenwei replied:
Fixed

katbyte commented on azurerm/resource_arm_netapp_account.go:
```diff
@@ -0,0 +1,288 @@
+			"active_directory": {
+				Type:     schema.TypeList,
+				Optional: true,
+				Elem: &schema.Resource{
```
Are we sure there can be more then one of these?
```
------- Stdout: -------
=== RUN   TestAccAzureRMNetAppAccount_update
=== PAUSE TestAccAzureRMNetAppAccount_update
=== CONT  TestAccAzureRMNetAppAccount_update
--- FAIL: TestAccAzureRMNetAppAccount_update (83.58s)
    testing.go:569: Step 1 error: errors during apply:
        
        Error: Error creating NetApp Account "acctest-NetAppAccount-191008004136081477" (Resource Group "acctestRG-191008004136081477"): netapp.AccountsClient#CreateOrUpdate: Failure sending request: StatusCode=400 -- Original Error: Code="ErrorUpsertingActiveDirectory" Message="Only one active directory allowed"
        
          on /opt/teamcity-agent/temp/buildTmp/tf-test307630958/main.tf line 7:
          (source code not available)```

neil-yechenwei replied:
After checked, it only supports one active directory. So I added attribute "MaxItems: 1".
![image](https://user-images.githubusercontent.com/19754191/66380832-bb916d80-e9ea-11e9-9844-9c72d7b291de.png)

neil-yechenwei commented on azurerm/resource_arm_netapp_account.go:
```diff
@@ -0,0 +1,288 @@
+		for k, v := range tagMap {
+			currentTags[k] = utils.String(v.(string))
+		}
+	}
```
It seems that we cannot do this since the type of resp.Tags is "Tags interface{} `json:"tags,omitempty"`" and function tags.FlattenAndSet only can accept type "tagMap map[string]*string". If I write code as "return tags.FlattenAndSet(d, resp.Tags)" and it would throw below error message.
![image](https://user-images.githubusercontent.com/19754191/66373687-cb558580-e9db-11e9-8565-0b08cbe58c5a.png)

katbyte replied:
tbf, this is probably an error/issue with the swagger and API and should be tracked down. The swagger is probably missing the type sois defaulting to an interface rather then the correct types

neil-yechenwei commented:
> Thanks for the revisions @neil-yechenwei! looking much better.
> 
> It looks like there can be only one `active_directory` block:
> 
> ```
> ------- Stdout: -------
> === RUN   TestAccAzureRMNetAppAccount_update
> === PAUSE TestAccAzureRMNetAppAccount_update
> === CONT  TestAccAzureRMNetAppAccount_update
> --- FAIL: TestAccAzureRMNetAppAccount_update (83.58s)
>     testing.go:569: Step 1 error: errors during apply:
>         
>         Error: Error creating NetApp Account "acctest-NetAppAccount-191008004136081477" (Resource Group "acctestRG-191008004136081477"): netapp.AccountsClient#CreateOrUpdate: Failure sending request: StatusCode=400 -- Original Error: Code="ErrorUpsertingActiveDirectory" Message="Only one active directory allowed"
>         
>           on /opt/teamcity-agent/temp/buildTmp/tf-test307630958/main.tf line 7:
>           (source code not available)
> ```

Thanks for the comments @katbyte . After checked for this resource in Azure portal, it only supports one active directory. So I added attribute "MaxItems: 1".
Azure Rest API doc for NetApp Account:
![image](https://user-images.githubusercontent.com/19754191/66444482-02787500-ea76-11e9-8a19-95d7c233d0c7.png)
NetApp Resource in Azure Portal:
![image](https://user-images.githubusercontent.com/19754191/66444509-13c18180-ea76-11e9-8b0e-6db24c5da730.png)

katbyte commented on azurerm/data_source_netapp_account.go:
```diff
@@ -0,0 +1,108 @@
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"password": {
```
This is returned blanked out so might not make sense in the data source?

neil-yechenwei replied:
FIxed

katbyte commented on website/docs/r/netapp_account.html.markdown:
```diff
@@ -0,0 +1,96 @@
+
+The `active_directory` block supports the following:
+
+* `id` - The resource id of Active Directory.
```
```suggestion
* `id` - The resource ID of Active Directory.
```

neil-yechenwei replied:
Fixed

neil-yechenwei commented:
Thanks for the comments, @katbyte ,
After reran this test case,  it passed on my local machine and travis CI. It seems that the issue you showed only happened on your test environment. The below warning shows "**Only one active directory can be joined to a single subscription**". So maybe the subscription what you used had joined the active directory. Could you remove all active directory connections under netapp account and then re-run this test case?

![image](https://user-images.githubusercontent.com/19754191/66459882-6ddb3a80-eaa8-11e9-8c2e-45049977bbbc.png)

My test result:
$ make testacc TESTARGS="-run=TestAccAzureRMNetAppAccount_update"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test $(go list ./... |grep -v 'vendor'|grep -v 'examples') -v -run=TestAccAzureRMNetAppAccount_update -timeout 180m -ldflags="-X=github.com/terraform-providers/terraform-provider-azurerm/version.ProviderVersion=acc"
?       github.com/terraform-providers/terraform-provider-azurerm       [no test files]
=== RUN   TestAccAzureRMNetAppAccount_update
=== PAUSE TestAccAzureRMNetAppAccount_update
=== CONT  TestAccAzureRMNetAppAccount_update
--- PASS: TestAccAzureRMNetAppAccount_update (159.20s)

neil-yechenwei commented:
Thanks for the comments, @katbyte ,
After reran this test case "TestAccAzureRMNetAppAccount_update",  it passed on my local machine and travis CI. It seems that the issue you showed only happened on your test environment. The below warning shows "**Only one active directory can be joined to a single subscription**". So maybe the subscription what you used to create netapp account resource had joined the active directory. Could you firstly remove all active directory connections under all netapp accounts from Azure side and then re-run this test case?

![image](https://user-images.githubusercontent.com/19754191/66459882-6ddb3a80-eaa8-11e9-8c2e-45049977bbbc.png)

My test result:
$ make testacc TESTARGS="-run=TestAccAzureRMNetAppAccount_update"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test $(go list ./... |grep -v 'vendor'|grep -v 'examples') -v -run=TestAccAzureRMNetAppAccount_update -timeout 180m -ldflags="-X=github.com/terraform-providers/terraform-provider-azurerm/version.ProviderVersion=acc"
?       github.com/terraform-providers/terraform-provider-azurerm       [no test files]
=== RUN   TestAccAzureRMNetAppAccount_update
=== PAUSE TestAccAzureRMNetAppAccount_update
=== CONT  TestAccAzureRMNetAppAccount_update
--- PASS: TestAccAzureRMNetAppAccount_update (159.20s)

neil-yechenwei commented:
Thanks for your suggestion for return type of resp.Tags. I think the return type of resp.Tags is "interface{}" not "map[string]*string" is fine. Maybe you're right but I am not sure why do you think it shouldn't be type "interface{}". Could you provide the justification for it? If it's necessary, I can file an issue to repo "Azure/azure-rest-api-specs" based on your justification. Before they process or fix this issue, could you merge this PR first for unblocking users to use this resource since we don't know when they can fix this issue?

katbyte commented:
Oh 100% that is not a blocker for this PR, but it's that every other resource it is `map[string]*string`,and that *is* what the tags property is, isn't it? There is no reason the netapp resource should be different then all the other resources and it adds unnecessary complexity.

Let me turn it around, what is the reason it should be `interface{}` and different then basically every other resource??

katbyte commented:
@neil-yechenwei, we don't have any netapp accounts currently:
![image](https://user-images.githubusercontent.com/1638467/66507086-66556a80-ea83-11e9-99e5-8e960edd4530.png)

Could it be that you can't run both tests at the same time? what happens if you run both locally in parallel?

neil-yechenwei commented:
It seems that I didn't run both tests at the same time. Generally, I run below test commands one by one on my local machine and they didn't throw any error. I think maybe it would throw error if run both tests at the same time. Because test cases "TestAccAzureRMNetAppAccount_complete" and "TestAccAzureRMNetAppAccount_update" would add same subscription to the active directory when run these test cases parallelly.  Could you rerun the test cases one by one on your local machine not parallel?
Test steps:
make testacc TESTARGS="-run=TestAccAzureRMNetAppAccount_basic"
make testacc TESTARGS="-run=TestAccAzureRMNetAppAccount_complete"
make testacc TESTARGS="-run=TestAccAzureRMNetAppAccount_update"
make testacc TESTARGS="-run=TestAccDataSourceAzureRMNetAppAccount_basic"

neil-yechenwei commented:
I think your concern is reasonable. I will file an issue to repo "Azure/azure-rest-api-specs" based on your justification. Before they process or fix this issue, please help to merge this PR first for unblocking users to use this resource since we don't know when they can fix this issue. Thanks.

katbyte commented:
Also could you please include that link to the rest specs issue in the code next to the code handling the bug? something along the lines of

`// handle tags being interface{} until https://github.com/Azure/azure-rest-api-specs/issues/7447 is fixed

neil-yechenwei commented:
@katbyte , Could you share the command to run test cases in parallel?

neil-yechenwei commented:
> Also could you please include that link to the rest specs issue in the code next to the code handling the bug? something along the lines of
> 
> `// handle tags being interface{} until [Azure/azure-rest-api-specs#7447](https://github.com/Azure/azure-rest-api-specs/issues/7447) is fixed

Sure. Added to the code.

katbyte commented:
@neil-yechenwei, 

`make fmt; make testacc TEST=./azuread TESTARGS="-parallel 11 -test.run=TestAccWhatever"` will run them at the same time

neil-yechenwei commented:
> @neil-yechenwei, our CI system runs all tests in parallel, if only one netapp account can be joined to a domain at a time? Please confirm this and then we need to document it in the docs & setup the tests to run one at a time. Please see the `TestAccAzureRMNetworkWatcher` test for an example of how to do this.

@katbyte , the error message 'Code="ErrorUpsertingActiveDirectory" Message="Only one active directory allowed"' you encountered and the warning "Only one active directory can be joined to a single subscription" from Azure portal which is returned from NetApp Account API what service owner developed prove that only one subscription can be joined to a domain at a time. And I've updated code and added description to document for this issue.

My test result:
$ make testacc TESTARGS="-run=TestAccAzureRMNetAppAccount"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test $(go list ./... |grep -v 'vendor'|grep -v 'examples') -v -run=TestAccAzureRMNetAppAccount -timeout 180m -ldflags="-X=github.com/terraform-providers/terraform-provider-azurerm/version.ProviderVersion=acc"
?       github.com/terraform-providers/terraform-provider-azurerm       [no test files]
=== RUN   TestAccAzureRMNetAppAccount
=== RUN   TestAccAzureRMNetAppAccount/basic
=== RUN   TestAccAzureRMNetAppAccount/basic/basic
=== PAUSE TestAccAzureRMNetAppAccount/basic/basic
=== CONT  TestAccAzureRMNetAppAccount/basic/basic
=== RUN   TestAccAzureRMNetAppAccount/complete
=== RUN   TestAccAzureRMNetAppAccount/complete/complete
=== PAUSE TestAccAzureRMNetAppAccount/complete/complete
=== CONT  TestAccAzureRMNetAppAccount/complete/complete
=== RUN   TestAccAzureRMNetAppAccount/update
=== RUN   TestAccAzureRMNetAppAccount/update/update
=== PAUSE TestAccAzureRMNetAppAccount/update/update
=== CONT  TestAccAzureRMNetAppAccount/update/update
=== RUN   TestAccAzureRMNetAppAccount/DataSource
=== RUN   TestAccAzureRMNetAppAccount/DataSource/basic
=== PAUSE TestAccAzureRMNetAppAccount/DataSource/basic
=== CONT  TestAccAzureRMNetAppAccount/DataSource/basic
--- PASS: TestAccAzureRMNetAppAccount (472.07s)
    --- PASS: TestAccAzureRMNetAppAccount/basic (0.00s)
        --- PASS: TestAccAzureRMNetAppAccount/basic/basic (117.25s)
    --- PASS: TestAccAzureRMNetAppAccount/complete (0.00s)
        --- PASS: TestAccAzureRMNetAppAccount/complete/complete (114.08s)
    --- PASS: TestAccAzureRMNetAppAccount/update (0.00s)
        --- PASS: TestAccAzureRMNetAppAccount/update/update (121.57s)
    --- PASS: TestAccAzureRMNetAppAccount/DataSource (0.00s)
        --- PASS: TestAccAzureRMNetAppAccount/DataSource/basic (119.18s)

neil-yechenwei commented:
> @neil-yechenwei,
> 
> `make fmt; make testacc TEST=./azuread TESTARGS="-parallel 11 -test.run=TestAccWhatever"` will run them at the same time

@katbyte , I've updated test cases and they can pass now.

katbyte commented on azurerm/resource_arm_netapp_account.go:
```diff
@@ -0,0 +1,280 @@
+							Optional: true,
+							Computed: true,
+						},
+						"id": {
```
This is also not in the docs?

katbyte replied:
or data source

neil-yechenwei replied:
It means the id of Active Directory, seems that it isn't used by other places so that I removed this field from resource file and document file.

neil-yechenwei commented on azurerm/resource_arm_netapp_account.go:
```diff
@@ -0,0 +1,280 @@
+							DiffSuppressFunc: func(k, old, new string, d *schema.ResourceData) bool {
+								maskedNew := strings.Replace(new, new, "****************", -1)
+								return (new == d.Get(k).(string)) && (maskedNew == old)
+							},
```
@katbyte , I am not sure what you mean about "we should pull the value forward from config/state". Could you share some code sample for your suggestion for this? I know "make test" would fail if I removed this diff suppress function.

$ make testacc TEST=./azurerm TESTARGS="-parallel 11 -test.run=TestAccAzureRMNetApp"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./azurerm -v -parallel 11 -test.run=TestAccAzureRMNetApp -timeout 180m -ldflags="-X=github.com/terraform-providers/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccAzureRMNetAppAccount
=== RUN   TestAccAzureRMNetAppAccount/update
=== RUN   TestAccAzureRMNetAppAccount/update/update
=== PAUSE TestAccAzureRMNetAppAccount/update/update
=== CONT  TestAccAzureRMNetAppAccount/update/update
=== RUN   TestAccAzureRMNetAppAccount/DataSource
=== RUN   TestAccAzureRMNetAppAccount/DataSource/basic
=== PAUSE TestAccAzureRMNetAppAccount/DataSource/basic
=== CONT  TestAccAzureRMNetAppAccount/DataSource/basic
=== RUN   TestAccAzureRMNetAppAccount/basic
=== RUN   TestAccAzureRMNetAppAccount/basic/basic
=== PAUSE TestAccAzureRMNetAppAccount/basic/basic
=== CONT  TestAccAzureRMNetAppAccount/basic/basic
=== RUN   TestAccAzureRMNetAppAccount/complete
=== RUN   TestAccAzureRMNetAppAccount/complete/complete
=== PAUSE TestAccAzureRMNetAppAccount/complete/complete
=== CONT  TestAccAzureRMNetAppAccount/complete/complete
--- FAIL: TestAccAzureRMNetAppAccount (436.81s)
    --- FAIL: TestAccAzureRMNetAppAccount/update (0.00s)
        --- FAIL: TestAccAzureRMNetAppAccount/update/update (115.72s)
            testing.go:569: Step 1 error: After applying this step, the plan was not empty:

                DIFF:

                UPDATE: azurerm_netapp_account.test
                  active_directory.#:                     "1" => "1"
                  active_directory.0.dns_servers.#:       "1" => "1"
                  active_directory.0.dns_servers.0:       "1.2.3.4" => "1.2.3.4"
                  active_directory.0.domain:              "westcentralus.com" => "westcentralus.com"
                  active_directory.0.id:                  "e6b944e9-fd12-4e92-12be-615da79a0137" => "e6b944e9-fd12-4e92-12be-615da79a0137"
                  active_directory.0.organizational_unit: "OU=FirstLevel" => "OU=FirstLevel"
                  active_directory.0.password:            "****************" => "aduserpwd"
                  active_directory.0.smb_server_name:     "SMBSERVER" => "SMBSERVER"
                  active_directory.0.username:            "aduser" => "aduser"
                  id:                                     "/subscriptions/67a9759d-d099-4aa8-8675-e6cfd669c3f4/resourceGroups/acctestRG-netapp-191019070306264354/providers/Microsoft.NetApp/netAppAccounts/acctest-NetAppAccount-191019070306264354" => "/subscriptions/67a9759d-d099-4aa8-8675-e6cfd669c3f4/resourceGroups/acctestRG-netapp-191019070306264354/providers/Microsoft.NetApp/netAppAccounts/acctest-NetAppAccount-191019070306264354"
                  location:                               "eastus2" => "eastus2"
                  name:                                   "acctest-NetAppAccount-191019070306264354" => "acctest-NetAppAccount-191019070306264354"
                  resource_group_name:                    "acctestRG-netapp-191019070306264354" => "acctestRG-netapp-191019070306264354"
                  tags.ENV:                               "Test" => "Test"



                STATE:

                azurerm_netapp_account.test:
                  ID = /subscriptions/67a9759d-d099-4aa8-8675-e6cfd669c3f4/resourceGroups/acctestRG-netapp-191019070306264354/providers/Microsoft.NetApp/netAppAccounts/acctest-NetAppAccount-191019070306264354
                  provider = provider.azurerm
                  active_directory.# = 1
                  active_directory.0.dns_servers.# = 1
                  active_directory.0.dns_servers.0 = 1.2.3.4
                  active_directory.0.domain = westcentralus.com
                  active_directory.0.id = e6b944e9-fd12-4e92-12be-615da79a0137
                  active_directory.0.organizational_unit = OU=FirstLevel
                  active_directory.0.password = ****************
                  active_directory.0.smb_server_name = SMBSERVER
                  active_directory.0.username = aduser
                  location = eastus2
                  name = acctest-NetAppAccount-191019070306264354
                  resource_group_name = acctestRG-netapp-191019070306264354
                  tags.ENV = Test

                  Dependencies:
                    azurerm_resource_group.test
                azurerm_resource_group.test:
                  ID = /subscriptions/67a9759d-d099-4aa8-8675-e6cfd669c3f4/resourceGroups/acctestRG-netapp-191019070306264354
                  provider = provider.azurerm
                  location = eastus2
                  name = acctestRG-netapp-191019070306264354
    --- PASS: TestAccAzureRMNetAppAccount/DataSource (0.00s)
        --- PASS: TestAccAzureRMNetAppAccount/DataSource/basic (103.02s)
    --- PASS: TestAccAzureRMNetAppAccount/basic (0.00s)
        --- PASS: TestAccAzureRMNetAppAccount/basic/basic (111.85s)
    --- FAIL: TestAccAzureRMNetAppAccount/complete (0.00s)
        --- FAIL: TestAccAzureRMNetAppAccount/complete/complete (106.22s)
            testing.go:569: Step 0 error: After applying this step, the plan was not empty:

                DIFF:

                UPDATE: azurerm_netapp_account.test
                  active_directory.#:                     "1" => "1"
                  active_directory.0.dns_servers.#:       "1" => "1"
                  active_directory.0.dns_servers.0:       "1.2.3.4" => "1.2.3.4"
                  active_directory.0.domain:              "westcentralus.com" => "westcentralus.com"
                  active_directory.0.id:                  "c116d2fa-297e-a04f-5e7d-97671e2f2ac9" => "c116d2fa-297e-a04f-5e7d-97671e2f2ac9"
                  active_directory.0.organizational_unit: "OU=FirstLevel" => "OU=FirstLevel"
                  active_directory.0.password:            "****************" => "aduserpwd"
                  active_directory.0.smb_server_name:     "SMBSERVER" => "SMBSERVER"
                  active_directory.0.username:            "aduser" => "aduser"
                  id:                                     "/subscriptions/67a9759d-d099-4aa8-8675-e6cfd669c3f4/resourceGroups/acctestRG-netapp-191019070836852308/providers/Microsoft.NetApp/netAppAccounts/acctest-NetAppAccount-191019070836852308" => "/subscriptions/67a9759d-d099-4aa8-8675-e6cfd669c3f4/resourceGroups/acctestRG-netapp-191019070836852308/providers/Microsoft.NetApp/netAppAccounts/acctest-NetAppAccount-191019070836852308"
                  location:                               "eastus2" => "eastus2"
                  name:                                   "acctest-NetAppAccount-191019070836852308" => "acctest-NetAppAccount-191019070836852308"
                  resource_group_name:                    "acctestRG-netapp-191019070836852308" => "acctestRG-netapp-191019070836852308"
                  tags.ENV:                               "Test" => "Test"



                STATE:

                azurerm_netapp_account.test:
                  ID = /subscriptions/67a9759d-d099-4aa8-8675-e6cfd669c3f4/resourceGroups/acctestRG-netapp-191019070836852308/providers/Microsoft.NetApp/netAppAccounts/acctest-NetAppAccount-191019070836852308
                  provider = provider.azurerm
                  active_directory.# = 1
                  active_directory.0.dns_servers.# = 1
                  active_directory.0.dns_servers.0 = 1.2.3.4
                  active_directory.0.domain = westcentralus.com
                  active_directory.0.id = c116d2fa-297e-a04f-5e7d-97671e2f2ac9
                  active_directory.0.organizational_unit = OU=FirstLevel
                  active_directory.0.password = ****************
                  active_directory.0.smb_server_name = SMBSERVER
                  active_directory.0.username = aduser
                  location = eastus2
                  name = acctest-NetAppAccount-191019070836852308
                  resource_group_name = acctestRG-netapp-191019070836852308
                  tags.ENV = Test

                  Dependencies:
                    azurerm_resource_group.test
                azurerm_resource_group.test:
                  ID = /subscriptions/67a9759d-d099-4aa8-8675-e6cfd669c3f4/resourceGroups/acctestRG-netapp-191019070836852308
                  provider = provider.azurerm
                  location = eastus2
                  name = acctestRG-netapp-191019070836852308
FAIL
FAIL    github.com/terraform-providers/terraform-provider-azurerm/azurerm       448.933s
make: *** [testacc] Error 1

katbyte replied:
Please don't resolve resolve conversations until a consensus has been reached.

What i mean here is that instead of doing a diff suppress we should on read grab the old value from state/config and reapply it instead of reading it from the API (because it will always be ***)

neil-yechenwei commented on azurerm/resource_arm_netapp_account_test.go:
```diff
@@ -0,0 +1,222 @@
+		},
+		"DataSource": {
+			"basic": testAccDataSourceAzureRMNetAppAccount_basic,
+		},
```
Actually, "make test" would still run in parallel and still throw same error "Error: Code="ErrorUpsertingActiveDirectory" Message="Only one active directory allowed"" you previously encountered if I use the code you suggested above.

$ make testacc TEST=./azurerm TESTARGS="-parallel 11 -test.run=TestAccAzureRMNetApp"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./azurerm -v -parallel 11 -test.run=TestAccAzureRMNetApp -timeout 180m -ldflags="-X=github.com/terraform-providers/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccAzureRMNetAppAccount
=== RUN   TestAccAzureRMNetAppAccount/resource
=== RUN   TestAccAzureRMNetAppAccount/resource/basic
=== PAUSE TestAccAzureRMNetAppAccount/resource/basic
=== RUN   TestAccAzureRMNetAppAccount/resource/complete
=== PAUSE TestAccAzureRMNetAppAccount/resource/complete
=== RUN   TestAccAzureRMNetAppAccount/resource/update
=== PAUSE TestAccAzureRMNetAppAccount/resource/update
=== CONT  TestAccAzureRMNetAppAccount/resource/basic
=== CONT  TestAccAzureRMNetAppAccount/resource/update
=== CONT  TestAccAzureRMNetAppAccount/resource/complete
=== RUN   TestAccAzureRMNetAppAccount/datasource
=== RUN   TestAccAzureRMNetAppAccount/datasource/basic
=== PAUSE TestAccAzureRMNetAppAccount/datasource/basic
=== CONT  TestAccAzureRMNetAppAccount/datasource/basic
--- FAIL: TestAccAzureRMNetAppAccount (219.82s)
    --- FAIL: TestAccAzureRMNetAppAccount/resource (0.01s)
        --- PASS: TestAccAzureRMNetAppAccount/resource/complete (98.19s)
        --- PASS: TestAccAzureRMNetAppAccount/resource/basic (113.31s)
        --- FAIL: TestAccAzureRMNetAppAccount/resource/update (117.87s)
            testing.go:569: Step 1 error: errors during apply:

                Error: Error creating NetApp Account "acctest-NetAppAccount-191019072148636553" (Resource Group "acctestRG-netapp-191019072148636553"): netapp.AccountsClient#CreateOrUpdate: Failure sending request: StatusCode=400 -- Original Error: Code="ErrorUpsertingActiveDirectory" Message="Only one active directory allowed"

                  on C:\Users\v-cheye\AppData\Local\Temp\tf-test227059795\main.tf line 7:
                  (source code not available)


    --- PASS: TestAccAzureRMNetAppAccount/datasource (0.00s)
        --- PASS: TestAccAzureRMNetAppAccount/datasource/basic (101.94s)
FAIL
FAIL    github.com/terraform-providers/terraform-provider-azurerm/azurerm       231.856s
make: *** [testacc] Error 1

katbyte replied:
then you need to change the loop logic to not run each of the categories at once

neil-yechenwei commented on website/docs/r/netapp_account.html.markdown:
```diff
@@ -3,7 +3,7 @@ layout: "azurerm"
 sidebar_current: "docs-azurerm-resource-netapp-account"
 description: |-
-  Manages Azure NetApp Account instance.
+  Manages Azure NetApp Account instance. Azure allows only one active directory can be joined to a single subscription at a time for NetApp Account.
```
Fixed

neil-yechenwei commented:
Thanks for the comments@katbyte ,

I've updated the code and comments. Thanks.

katbyte commented on azurerm/internal/services/netapp/validate.go:
```diff
@@ -0,0 +1,42 @@
+	}
+
+	return nil, errors
+}
```
There is an existing ipv4 validation function that we can use in the `validate` package

neil-yechenwei replied:
Fixed

katbyte commented on azurerm/resource_arm_netapp_account.go:
```diff
@@ -0,0 +1,273 @@
+							Required: true,
+							Elem: &schema.Schema{
+								Type:         schema.TypeString,
+								ValidateFunc: aznetapp.ValidateActiveDirectoryDNSName,
```
Does this only support ipv4 values? i would have assumed a FQDN would work as well. `dns.hashicorp.com`?

If it is only IPv4 addresses that are allowed could we use the existing IPv4 validator?

neil-yechenwei replied:
After checked, it doesn't support IPv6 and FQDN. So I've updated code to use the existing IPv4 validator. Thanks.
![image](https://user-images.githubusercontent.com/19754191/67255284-de745500-f4b3-11e9-9244-14d5c72a3f8f.png)
![image](https://user-images.githubusercontent.com/19754191/67255287-e2a07280-f4b3-11e9-9c41-46a5dadca52c.png)

neil-yechenwei replied:
After confirmed with the service owner, they also confirmed only IPv4 is permitted.

katbyte commented on azurerm/resource_arm_netapp_account.go:
```diff
@@ -0,0 +1,273 @@
+		}
+		if v := item.Password; v != nil {
+			b["password"] = *v
+		}
```
here, what we should be doing is something along the lines of:
```
if v, ok := d.GetOk("active_directory.0.password"); ok {
    b["password"] = v
}
```

neil-yechenwei replied:
Fixed

katbyte commented on azurerm/resource_arm_netapp_account_test.go:
```diff
@@ -0,0 +1,222 @@
+			}
+		})
+	}
+}
```
this should instead be something like:
```
    for group, m := range testCases {
			for name, tc := range m {
				tc := tc
				t.Run(name, func(t *testing.T) {
					tc(t)
				})
			}
		})
	}
```

neil-yechenwei replied:
Fixed

katbyte commented on website/docs/d/netapp_account.html.markdown:
```diff
@@ -0,0 +1,59 @@
+# Data Source: azurerm_netapp_account
+
+Uses this data source to access information about an existing NetApp Account.
+
```
```suggestion
```

neil-yechenwei replied:
Fixed

neil-yechenwei commented:
Hi @katbyte ,

Thank you for the comments, I've updated code, thanks.

tombuildsstuff commented on website/docs/d/netapp_account.html.markdown:
```diff
@@ -0,0 +1,58 @@
+
+* `location` - Specifies the supported Azure location where the resource exists. Changing this forces a new resource to be created.
+
+* `active_directory` - One or more `active_directory` block defined below.
```
there'll only be a single one of these:

```suggestion
* `active_directory` - An `active_directory` block defined below.
```

neil-yechenwei replied:
FIxed

tombuildsstuff commented on website/docs/r/netapp_account.html.markdown:
```diff
@@ -0,0 +1,89 @@
+page_title: "Azure Resource Manager: azurerm_netapp_account"
+sidebar_current: "docs-azurerm-resource-netapp-account"
+description: |-
+  Manages Azure NetApp Account instance.
```
```suggestion
  Manages a NetApp Account.
```

neil-yechenwei replied:
FIxed

tombuildsstuff commented on website/docs/r/netapp_account.html.markdown:
```diff
@@ -0,0 +1,89 @@
+}
+
+resource "azurerm_netapp_account" "example" {
+  name                = "acctestnetappaccount"
```
```suggestion
  name                = "example-netapp"
```

neil-yechenwei replied:
FIxed

tombuildsstuff commented on website/docs/r/netapp_account.html.markdown:
```diff
@@ -0,0 +1,89 @@
+
+* `domain` - (Required) Name of the Active Directory domain.
+
+* `smb_server_name` - (Required) NetBIOS name of the SMB server. This name will be registered as a computer account in the AD and used to mount volumes.
```
I think this'd be clearer as:

```suggestion
* `smb_server_name` - (Required) The NetBIOS name which should be used for the NetApp SMB Server, which will be registered as a computer account in the AD and used to mount volumes.
```

neil-yechenwei replied:
Fixed

tombuildsstuff commented on website/docs/r/netapp_account.html.markdown:
```diff
@@ -0,0 +1,89 @@
+
+* `smb_server_name` - (Required) NetBIOS name of the SMB server. This name will be registered as a computer account in the AD and used to mount volumes.
+
+* `username` - (Required) Username of Active Directory domain administrator, which have permissions to create a SMB machine account in the AD domain.
```
```suggestion
* `username` - (Required) The Username of Active Directory Domain Administrator.

-> **NOTE:** This account will be used to create the Machine Account within the Active Directory Domain
```

neil-yechenwei replied:
FIxed

tombuildsstuff commented on website/docs/r/netapp_account.html.markdown:
```diff
@@ -0,0 +1,89 @@
+
+* `password` - (Required) Plain text password of Active Directory domain administrator.
+
+* `organizational_unit` - (Optional) The Organizational Unit (OU) within the Windows Active Directory.
```
```suggestion
* `organizational_unit` - (Optional) The Organizational Unit (OU) within the Active Directory Domain.
```

neil-yechenwei replied:
Fixed

tombuildsstuff commented on website/docs/r/netapp_account.html.markdown:
```diff
@@ -0,0 +1,89 @@
+
+The following attributes are exported:
+
+* `id` - The ID of the NetApp resource.
```
```suggestion
* `id` - The ID of the NetApp Account.
```

neil-yechenwei replied:
FIxed

tombuildsstuff commented on website/docs/r/netapp_account.html.markdown:
```diff
@@ -0,0 +1,89 @@
+
+## Import
+
+NetApp Account can be imported using the `resource id`, e.g.
```
```suggestion
NetApp Accounts can be imported using the `resource id`, e.g.
```

neil-yechenwei replied:
Fixed

tombuildsstuff commented on website/docs/r/netapp_account.html.markdown:
````diff
@@ -0,0 +1,89 @@
+NetApp Account can be imported using the `resource id`, e.g.
+
+```shell
+$ terraform import azurerm_netapp_account.example /subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/acctestRG/providers/Microsoft.NetApp/netAppAccounts/acctestnetappaccount
````
to match other resources can we make this:

```suggestion
$ terraform import azurerm_netapp_account.example /subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/group1/providers/Microsoft.NetApp/netAppAccounts/account1
```

neil-yechenwei replied:
fixed

tombuildsstuff commented on azurerm/resource_arm_netapp_account_test.go:
```diff
@@ -0,0 +1,224 @@
+	testCases := map[string]map[string]func(t *testing.T){
+		"Resource": {
+			"basic":    testAccAzureRMNetAppAccount_basic,
+			"complete": testAccAzureRMNetAppAccount_complete,
```
we also need to add a Requires Import test here - example: https://github.com/terraform-providers/terraform-provider-azurerm/blob/master/azurerm/resource_arm_resource_group_test.go#L38-L64

neil-yechenwei replied:
Fixed

tombuildsstuff commented on azurerm/resource_arm_netapp_account.go:
```diff
@@ -0,0 +1,269 @@
+				ValidateFunc: aznetapp.ValidateNetAppAccountName,
+			},
+
+			"resource_group_name": azure.SchemaResourceGroupNameForDataSource(),
```
this should be:

```suggestion
			"resource_group_name": azure.SchemaResourceGroupName(),
```

neil-yechenwei replied:
fixed

neil-yechenwei commented on azurerm/data_source_netapp_account.go:
```diff
@@ -0,0 +1,104 @@
+	}
+
+	// Handles tags being interface{} until https://github.com/Azure/azure-rest-api-specs/issues/7447 is fixed
+	currentTags := make(map[string]*string)
```
@tombuildsstuff , so currently I should remove tags from the schema list, right?

tombuildsstuff replied:
yep; once https://github.com/Azure/azure-rest-api-specs/issues/7447 is resolved we can add these in

neil-yechenwei replied:
Fixed

neil-yechenwei commented on azurerm/internal/services/netapp/validate.go:
```diff
@@ -0,0 +1,27 @@
+)
+
+func ValidateNetAppAccountName(i interface{}, k string) (_ []string, errors []error) {
+	if m, regexErrs := validate.RegExHelper(i, k, `(^[\da-zA-Z])([-\da-zA-Z]{1,62})([\da-zA-Z]$)`); !m {
```
@tombuildsstuff , Do you mean I need to directly use regex validator "regexp" here not using "validate.RegExHelper", right?

katbyte replied:
We have a `validation.StringMatch` helper that let you pass in regex to match against.

katbyte replied:
we have the validation function `validation.StringMatch` we can use instead of rolling our own

neil-yechenwei replied:
fixed

neil-yechenwei commented on azurerm/data_source_netapp_account.go:
```diff
@@ -0,0 +1,104 @@
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/azure"
+	aznetapp "github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/netapp"
```
Fixed

neil-yechenwei commented on azurerm/internal/services/netapp/client.go:
```diff
@@ -0,0 +1,19 @@
+}
+
+func BuildClient(o *common.ClientOptions) *Client {
+	AccountClient := netapp.NewAccountsClientWithBaseURI(o.ResourceManagerEndpoint, o.SubscriptionId)
```
Fixed

neil-yechenwei commented on azurerm/resource_arm_netapp_account.go:
```diff
@@ -0,0 +1,269 @@
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/tf"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/validate"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/features"
+	aznetapp "github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/netapp"
```
Fixed

neil-yechenwei commented on azurerm/resource_arm_netapp_account.go:
```diff
@@ -0,0 +1,269 @@
+		for k, v := range tagMap {
+			currentTags[k] = utils.String(v.(string))
+		}
+	}
```
Removed

neil-yechenwei commented on azurerm/resource_arm_netapp_account.go:
```diff
@@ -0,0 +1,269 @@
+	if err != nil {
+		if response.WasNotFound(future.Response()) {
+			return nil
+		}
```
Fixed

neil-yechenwei commented on examples/netapp/account/variables.tf:
```diff
@@ -0,0 +1,23 @@
+
+variable "location" {
+  description = "The Azure location where all resources in this example should be created."
+  default     = "eastus2"
```
Fixed

neil-yechenwei commented on examples/netapp/account/main.tf:
```diff
@@ -0,0 +1,23 @@
+resource "azurerm_resource_group" "example" {
+  name     = "${var.resource_group_name}"
```
Fixed

neil-yechenwei commented on examples/netapp/account/variables.tf:
```diff
@@ -0,0 +1,23 @@
+variable "resource_group_name" {
+  description = "The name of the resource group the NetApp Account is located in."
+}
```
Fixed

neil-yechenwei commented on examples/netapp/account/variables.tf:
```diff
@@ -0,0 +1,23 @@
+
+variable "prefix" {
+  description = "The prefix used for all resources used by this NetApp Account"
+  default     = "example"
```
Fixed

neil-yechenwei commented on examples/netapp/account/main.tf:
```diff
@@ -0,0 +1,23 @@
+
+  active_directory {
+    username            = "${var.username}"
+    password            = "${var.password}"
```
Fixed

neil-yechenwei commented on examples/netapp/account/variables.tf:
```diff
@@ -0,0 +1,23 @@
+variable "password" {
+  description = "The password of Active Directory"
+  default     = "aduserpwd"
+}
```
Fixed

neil-yechenwei commented on website/docs/d/netapp_account.html.markdown:
```diff
@@ -0,0 +1,58 @@
+
+The following attributes are exported:
+
+* `location` - Specifies the supported Azure location where the resource exists. Changing this forces a new resource to be created.
```
Fixed

neil-yechenwei commented on website/docs/d/netapp_account.html.markdown:
```diff
@@ -0,0 +1,58 @@
+
+* `active_directory` - One or more `active_directory` block defined below.
+
+* `tags` - A mapping of tags to assign to the resource.
```
Fixed

neil-yechenwei commented on website/docs/d/netapp_account.html.markdown:
```diff
@@ -0,0 +1,58 @@
+
+---
+
+The `active_directory` block contains the following:
```
Fixed

neil-yechenwei commented on website/docs/d/netapp_account.html.markdown:
```diff
@@ -0,0 +1,58 @@
+
+The `active_directory` block contains the following:
+
+* `dns_servers` - A list of DNS server IP addresses for the Active Directory domain.
```
Fixed

neil-yechenwei commented on website/docs/d/netapp_account.html.markdown:
```diff
@@ -0,0 +1,58 @@
+
+* `dns_servers` - A list of DNS server IP addresses for the Active Directory domain.
+
+* `domain` - Name of the Active Directory domain.
```
Fixed

neil-yechenwei commented on website/docs/d/netapp_account.html.markdown:
```diff
@@ -0,0 +1,58 @@
+
+* `domain` - Name of the Active Directory domain.
+
+* `smb_server_name` - NetBIOS name of the SMB server. This name will be registered as a computer account in the AD and used to mount volumes.
```
FIxed

neil-yechenwei commented on website/docs/d/netapp_account.html.markdown:
```diff
@@ -0,0 +1,58 @@
+
+* `smb_server_name` - NetBIOS name of the SMB server. This name will be registered as a computer account in the AD and used to mount volumes.
+
+* `username` - Username of Active Directory domain administrator, which have permissions to create a SMB machine account in the AD domain.
```
Fixed

neil-yechenwei commented on website/docs/d/netapp_account.html.markdown:
```diff
@@ -0,0 +1,58 @@
+
+* `username` - Username of Active Directory domain administrator, which have permissions to create a SMB machine account in the AD domain.
+
+* `organizational_unit` - The Organizational Unit (OU) within the Windows Active Directory Domain the netapp is placed.
```
Fixed

neil-yechenwei commented on website/docs/r/netapp_account.html.markdown:
```diff
@@ -0,0 +1,89 @@
+
+# azurerm_netapp_account
+
+Manages Azure NetApp Account instance.
```
Fixed

neil-yechenwei commented on website/docs/r/netapp_account.html.markdown:
````diff
@@ -0,0 +1,89 @@
+
+```hcl
+resource "azurerm_resource_group" "example" {
+  name     = "acctestRG"
````
Fixed

neil-yechenwei commented on website/docs/r/netapp_account.html.markdown:
````diff
@@ -0,0 +1,89 @@
+```hcl
+resource "azurerm_resource_group" "example" {
+  name     = "acctestRG"
+  location = "Eastus2"
````
Fixed

neil-yechenwei commented on website/docs/r/netapp_account.html.markdown:
```diff
@@ -0,0 +1,89 @@
+
+* `location` - (Required) Specifies the supported Azure location where the resource exists. Changing this forces a new resource to be created.
+
+* `active_directory` - (Optional) One or more `active_directory` block defined below.
```
Fixed

neil-yechenwei commented on website/docs/r/netapp_account.html.markdown:
```diff
@@ -0,0 +1,89 @@
+
+* `name` - (Required) The name of the NetApp Account. Changing this forces a new resource to be created.
+
+* `resource_group_name` - (Required) The name of the resource group where the NetApp Account resides. Changing this forces a new resource to be created.
```
FIxed

neil-yechenwei commented on website/docs/r/netapp_account.html.markdown:
```diff
@@ -0,0 +1,89 @@
+
+* `dns_servers` - (Required) A list of DNS server IP addresses for the Active Directory domain. Only allows `IPv4` address.
+
+* `domain` - (Required) Name of the Active Directory domain.
```
FIxed

neil-yechenwei commented on website/docs/r/netapp_account.html.markdown:
```diff
@@ -0,0 +1,89 @@
+
+* `username` - (Required) Username of Active Directory domain administrator, which have permissions to create a SMB machine account in the AD domain.
+
+* `password` - (Required) Plain text password of Active Directory domain administrator.
```
Fixed

neil-yechenwei commented on website/docs/r/netapp_account.html.markdown:
```diff
@@ -0,0 +1,89 @@
+
+Manages Azure NetApp Account instance.
+
+~> **NOTE:** Azure allows only one active directory can be joined to a single subscription at a time for NetApp Account.
```
After netapp account resource is created in azure side, only one active directory can be joined to this netapp account in one subscription.  No more information. No ETA for this. Below information is from the owner of this service. Maybe you should read context in this PR or ask katbyte to get more details.
![image](https://user-images.githubusercontent.com/19754191/67543527-65c7ff80-f724-11e9-9259-88b9be4b7f48.png)

neil-yechenwei commented on azurerm/resource_arm_netapp_account_test.go:
```diff
@@ -0,0 +1,224 @@
+    password            = "aduserpwd"
+    smb_server_name     = "SMBSERVER"
+    dns_servers         = ["1.2.3.4"]
+	domain              = "westcentralus.com"
```
No! After netapp account resource is created in azure side, only one active directory can be joined to this netapp account in one subscription. Maybe you should read context in this PR or ask katbyte to get more details.

katbyte replied:
confirmed this, when you try and set two different ones you get:
```
 Error: Error waiting for creation of NetApp Account "acctest-NetAppAccount-191114162218570150" (Resource Group "acctestRG-netapp-191114162218570150"): Code="InternalServerError" Message="Error creating ActiveDirectory."
```
🤷‍♀

neil-yechenwei commented on azurerm/resource_arm_netapp_account.go:
```diff
@@ -0,0 +1,269 @@
+		}
+		if v := item.OrganizationalUnit; v != nil {
+			b["organizational_unit"] = *v
+		}
```
These feilds except "OrganizationalUnit" are always set. Please see the already implemented feature in azure portal.
![image](https://user-images.githubusercontent.com/19754191/67543875-e4716c80-f725-11e9-85af-6fc7c8a3d99d.png)

katbyte replied:
That's not what he means, he says that we need to set the properties on the data source regardless of if the API returns anything, see his code example.

neil-yechenwei replied:
I think we need these to identify which AD has been joined.

neil-yechenwei replied:
By now, I just know other resources like netapp pool and etc don't rely on these fields. So I removed them.

neil-yechenwei commented on azurerm/resource_arm_netapp_account.go:
```diff
@@ -0,0 +1,269 @@
+						"smb_server_name": {
+							Type:         schema.TypeString,
+							Required:     true,
+							ValidateFunc: validate.NoEmptyStrings,
```
fixed

neil-yechenwei commented on azurerm/resource_arm_netapp_account.go:
```diff
@@ -0,0 +1,269 @@
+						"organizational_unit": {
+							Type:     schema.TypeString,
+							Optional: true,
+							Computed: true,
```
fixed

neil-yechenwei commented on azurerm/internal/services/netapp/validate.go:
```diff
@@ -0,0 +1,27 @@
+}
+
+func ValidateActiveDirectoryDomainName(i interface{}, k string) (_ []string, errors []error) {
+	if m, regexErrs := validate.RegExHelper(i, k, `(^(.*)[\da-zA-Z]\.[\da-zA-Z].{1,}$)`); !m {
```
I don't know what you mean. You mean I need to use "regexp" not using RegexHelper. Could you share sample code for your suggestion?

katbyte replied:
We have a `validation.StringMatch` helper that let you pass in regex to match against.

katbyte replied:
we have the validation function `validation.StringMatch` we can use instead of rolling our own

neil-yechenwei replied:
fixed

neil-yechenwei commented on azurerm/internal/services/netapp/validate.go:
```diff
@@ -0,0 +1,27 @@
+	return nil, errors
+}
+
+func ValidateActiveDirectoryDomainName(i interface{}, k string) (_ []string, errors []error) {
```
Fixed

neil-yechenwei commented on azurerm/internal/services/netapp/validate.go:
```diff
@@ -0,0 +1,27 @@
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/validate"
+)
+
+func ValidateNetAppAccountName(i interface{}, k string) (_ []string, errors []error) {
```
Fixed

neil-yechenwei commented:
@tombuildsstuff , I've updated code by your comments. Please have a look. Thanks.

katbyte commented on azurerm/resource_arm_netapp_account_test.go:
```diff
@@ -0,0 +1,376 @@
+`, rInt, location, rInt)
+}
+
+func testAccAzureRMNetAppAccount_withIncorrectAccountNameConfig(rInt int, location string) string {
```
We don't need a test config here, we can just test the validation function

neil-yechenwei replied:
I've used validation.stringmatch so that i removed the test function here.

katbyte commented on azurerm/resource_arm_netapp_account_test.go:
```diff
@@ -0,0 +1,376 @@
+`, rInt, location)
+}
+
+func testAccAzureRMNetAppAccount_withIncorrectDomainNameConfig(rInt int, location string) string {
```
We don't need a test config here, we can just test the validation function

neil-yechenwei replied:
I've used validation.stringmatch so that i removed the test function here.

katbyte commented on azurerm/resource_arm_netapp_account_test.go:
```diff
@@ -0,0 +1,376 @@
+`, rInt, location, rInt)
+}
+
+func testAccAzureRMNetAppAccount_withIncorrectSMBServerNameConfig(rInt int, location string) string {
```
We don't need a test config here, we can just test the validation function

neil-yechenwei replied:
I've used validation.stringmatch so that i removed the test function here.

neil-yechenwei commented:
@katbyte ,Thanks for your comments.  I've updated code by your comments.

jstewart612 commented:
We're using ANF a lot more and are eager for this feature.  Near as I can tell, @katbyte and @tombuildsstuff , @neil-yechenwei has implemented your latest round of changes from your reviews.  Anything else blocking?  I'm not an allowed reviewer, obviously.

neil-yechenwei commented:
@katbyte , I am not sure what you mean about "confirmed this, when you try and set two different ones you get:". Are you trying to test for changing the active directory from A to B, right? If yes, it seems that it works fine from my side.
![image](https://user-images.githubusercontent.com/19754191/68910357-73553000-078c-11ea-90eb-5e0551502a63.png)
![image](https://user-images.githubusercontent.com/19754191/68910364-7c460180-078c-11ea-92b1-70ea7b5afce4.png)
